cocaine/detail/secservice.py

```diff
@@ -0,0 +1,96 @@
+#
+# Mostly compy-pasted from cocaine-tools
+#
+from tornado import gen
+from tornado.ioloop import IOLoop
+
+from cocaine.services import Service
+from cocaine.exceptions import CocaineError
+
+
+class SecureServiceError(CocaineError):
+    pass
+
+
+class BasicSecurity(object):
+    pass
+
+
+class Promiscuous(BasicSecurity):
+    @gen.coroutine
+    def fetch_token(self):
+        raise gen.Return('')
+
+
+class TVM(BasicSecurity):
+
+    def __init__(self, client_id, client_secret, name='tvm'):
+        self._client_id = client_id
+        self._client_secret = client_secret
+
+        self._tvm = Service(name)
+
+    @classmethod
+    def ty(self):
+        return 'TVM'
+
+    @gen.coroutine
+    def fetch_token(self):
+        grant_type = 'client_credentials'
+
+        channel = yield self._tvm.ticket_full(self._client_id, self._client_secret, grant_type, {})
+        ticket = yield channel.rx.get()
+
+        raise gen.Return(self._make_header(ticket))
+
+    def _make_header(self, ticket):
+        return '{} {}'.format(self.ty(), ticket)
+
+
+class SecureService(object):
+
+    def __init__(self, wrapped, secure, tok_update_sec, loop=IOLoop.current()):
+        self._wrapped = wrapped
+        self._secure = secure
+
+        self._tok_update_sec = tok_update_sec
+
+        loop.spawn_callback( lambda: self._refresh_token())
```

Comment:
Seems that @3Hren is right and we should try to avoid async poll completely as it is not warranty about async token fetch completion before token will be requested.  As a bonus it should free us from busywork with ioloop stuff.